flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManager.java

```diff
@@ -0,0 +1,609 @@
+		TaskExecutorConnection taskExecutorConnection = taskManagerSlot.getTaskManagerConnection();
+		TaskExecutorGateway gateway = taskExecutorConnection.getTaskExecutorGateway();
+
+		final CompletableFuture<Acknowledge> completableFuture = new CompletableFuture<>();
+		final SlotID slotId = taskManagerSlot.getSlotId();
+		final InstanceID instanceId = taskManagerSlot.getInstanceId();
+
+		LOG.debug("Starting allocation of slot {} for job {} with resource profile {}.", taskManagerSlot.getSlotId(), jobId, resourceProfile);
+
+		slotTracker.notifyAllocationStart(slotId, jobId);
+
+		if (!taskExecutorManager.isTaskManagerRegistered(instanceId)) {
+			throw new IllegalStateException("Could not find a registered task manager for instance id " +
+				instanceId + '.');
+		}
+
+		taskExecutorManager.markUsed(instanceId);
+
+		// RPC call to the task manager
+		CompletableFuture<Acknowledge> requestFuture = gateway.requestSlot(
+			slotId,
+			jobId,
+			new AllocationID(),
+			resourceProfile,
+			targetAddress,
+			resourceManagerId,
+			taskManagerRequestTimeout);
+
+		requestFuture.whenComplete(
+			(Acknowledge acknowledge, Throwable throwable) -> {
+				if (acknowledge != null) {
+					completableFuture.complete(acknowledge);
+				} else {
+					completableFuture.completeExceptionally(throwable);
+				}
+			});
+
+		completableFuture.whenCompleteAsync(
+			(Acknowledge acknowledge, Throwable throwable) -> {
+				try {
+					if (acknowledge != null) {
+						LOG.trace("Completed allocation of slot {} for job {}.", slotId, jobId);
+						slotTracker.notifyAllocationComplete(slotId, jobId);
+					} else {
+						if (throwable instanceof SlotOccupiedException) {
+							SlotOccupiedException exception = (SlotOccupiedException) throwable;
+							LOG.debug("Tried allocating slot {} for job {}, but it was already allocated for job {}.", slotId, jobId, exception.getJobId());
+							// report as a slot status to force the state transition
+							// this could be a problem if we ever assume that the task executor always reports about all slots
+							slotTracker.notifySlotStatus(Collections.singleton(new SlotStatus(slotId, taskManagerSlot.getResourceProfile(), exception.getJobId(), exception.getAllocationId())));
+						} else {
+							if (throwable instanceof CancellationException) {
+								LOG.debug("Cancelled allocation of slot {} for job {}.", slotId, jobId, throwable);
+							} else {
+								LOG.warn("Slot allocation for slot {} for job {} failed.", slotId, jobId, throwable);
+								slotTracker.notifyFree(slotId);
+							}
+						}
+					}
+					checkResourceRequirements();
+				} catch (Exception e) {
+					LOG.error("Error while completing the slot allocation.", e);
+				}
+			},
+			mainThreadExecutor);
+	}
+
+	private void tryFulfillRequirementsWithPendingOrNewSlots(JobID jobId, Map<ResourceProfile, Integer> missingResources, ResourceCounter pendingSlots) {
```

Comment:
Aren't we calling this method always with pending slots and no new slots?

Reply:
yes

Reply:
Then let's change the name of the method.